src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -742,5 +756,84 @@ private void UpdateCacheExpirationTime(KeyValueWatcher changeWatcher, bool succe
         }
+
+        private async Task ExecuteWithFailOverPolicyAsync(Func<ConfigurationClient, Task> funcToExecute, CancellationToken cancellationToken = default)
+        {
+            using IEnumerator<ConfigurationClient> clientEnumerator = _configClientProvider.GetClients().GetEnumerator();
+
+            // Guaranteed to return at least 1 element
+            Debug.Assert(clientEnumerator.MoveNext());
+
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    await funcToExecute(clientEnumerator.Current);
+
+                    _configClientProvider.UpdateClientStatus(clientEnumerator.Current, successful: true);
+
+                    break;
+                }
+                catch (AggregateException ae)
+                {
+                    if (!ShouldFailoverForException(ae))
+                    {
+                        throw;
+                    }
+
+                    _configClientProvider.UpdateClientStatus(clientEnumerator.Current, successful: false);
+
+                    if (!clientEnumerator.MoveNext())
+                    {
+                        throw;
+                    }
+                }
+                catch (RequestFailedException e)
+                {
+                    if (!ShouldFailoverForException(e))
+                    {
+                        throw;
+                    }
+
+                    _configClientProvider.UpdateClientStatus(clientEnumerator.Current, successful: false);
+
+                    if (!clientEnumerator.MoveNext())
+                    {
+                        throw;
+                    }
+                }
+            }
+        }
+
+        private bool ShouldFailoverForException(Exception ex)
+        {
+            int statusCode = HttpStatusCodes.NoResponse;
+
+            if (ex is AggregateException aggregateException)
+            {
+                IReadOnlyCollection<Exception> innerExceptions = aggregateException.Flatten().InnerExceptions;
+
+                if (innerExceptions.Count > 0 && innerExceptions.All(ex => ex is RequestFailedException))
+                {
+                    statusCode = (innerExceptions.Last() as RequestFailedException).Status;
+                }
+            }
+            else if (ex is RequestFailedException rfe)
+            {
+                statusCode = rfe.Status;
+
+                // The InnerException could be SocketException or WebException when endpoint is invalid
+                if ((statusCode == HttpStatusCodes.NoResponse && ex.InnerException is HttpRequestException hre && hre.InnerException != null) &&
+                    (hre.InnerException is WebException we && we.Status == WebExceptionStatus.NameResolutionFailure ||
+                    (hre.InnerException is SocketException se && se.SocketErrorCode == SocketError.HostNotFound)))
+                {
+                    return true;

```

Comment:
Why is invalid endpoint failover-able? If even one endpoint is invalid, we will waste resources trying to reach the endpoint and it will always be in backoff state. Instead, we could just throw the error and let the user know that they have provided an invalid endpoint.

That makes me think if we should ping all endpoints at startup and fail early if any of them are invalid.

Reply:
The thought behind keeping them failover-able was that the DNS resolution errors can also be transient (when client configures incorrectly, or temporary throttling of DNS queries from the DNS server) etc.

But I hear your argument; The possibility of DNS resolution error being transient is quite low and it is quite high when user provides incorrect endpoints. I am in double mind.

I'd wait to see what @zhenlan and @jimmyca15 say.

Reply:
Does DNS throttling also throw the same error code as host not found? 

I do see another reason for keeping your current logic though. If this error occurs because of a misconfigured firewall rule, users can fix their firewall settings without affecting their app. Another similar scenario is users deleting and recreating a store with the same name after a few minutes/hours/days. While the store is deleted, all requests will result in HostNotFound error, but it will be fixed when the store is created again.
So in a way, this error can be called transient. 
Worst case scenario if we treat invalid endpoints as failoverable - one request will be made to the invalid endpoint at max every 10 minutes if all other endpoints are down.